tools/checked-c-convert/ProgramInfo.cpp

```diff
@@ -31,6 +31,183 @@ bool ProgramInfo::checkStructuralEquality(uint32_t V, uint32_t U) {
   return false;
 }
 
+bool ProgramInfo::link() {
+
+  // For every global symbol in all the global symbols that we have found
+  // go through and apply rules for whether they are functions or variables.
+
+  for (const auto &S : GlobalSymbols) {
+    // First, extract out the function symbols from S. 
+    std::set<GlobalFunctionSymbol*> funcs;
+    std::set<GlobalVariableSymbol*> vars;
+    for (const auto &U : S.second)
+      if (GlobalFunctionSymbol *K = dyn_cast<GlobalFunctionSymbol>(U))
+        funcs.insert(K);
+      else if (GlobalVariableSymbol *V = dyn_cast<GlobalVariableSymbol>(U))
+        vars.insert(V);
+
+    // Then, pairwise-iterate over each of the function symbols found for this 
+    // symbol. What we want to do is for a sequence of constraint variables on
+    // F1,F2, set the constraint variables for F1(v0,vi,vN) and F2(v0,vj,vN) to
+    // be equal to each other, i.e. to enter a series of constraints of the form
+    // vi == vj.
+    //
+    // For example, consider a function that has been forward declared named 
+    // d1 with the signature void d1(int **c); There will be two sets of constraint
+    // variables in the system, one at the site of forward declaration and one at
+    // the site of function definition, like this:
+    //
+    // void d1(int *q_0 * q_1 c);
+    //
+    // void d1(int *q_2 * q_3 c) { *c = 0; }
+    //
+    // What we want to do is set q_0 == q_2 and q_1 == q_3. 
+    // To do that, we need to get the constraints for each parmvar decl position
+    // individually and set them equal, pairwise. 
+    for (std::set<GlobalFunctionSymbol*>::iterator I = funcs.begin();
+      I != funcs.end(); ++I) {
+      std::set<GlobalFunctionSymbol*>::iterator J = I;
+      J++;
+      if (J != funcs.end()) {
+        std::set<uint32_t> &rVars1 = (*I)->getReturns();
+        std::set<uint32_t> &rVars2 = (*J)->getReturns();
+        if (rVars1.size() == rVars2.size()) {
+
+          for (std::set<uint32_t>::iterator V1 = rVars1.begin(), V2 = rVars2.begin();
+            V1 != rVars1.end() && V2 != rVars2.end(); ++V1, ++V2)
+            CS.addConstraint(CS.createEq(
+              CS.getOrCreateVar(*V1), CS.getOrCreateVar(*V2)));
+        } else {
+          // Nothing makes sense because this means that the types of two 
+          // functions with the same name is different. Constrain 
+          // everything to top.
```

Comment:
Probably worth creating a helper function for setting everything to top.   It's fine to do that in a separate commit.